activerecord/lib/active_record/core.rb

```diff
@@ -308,12 +308,22 @@ def relation
           relation = Relation.create(self, arel_table, predicate_builder)
 
           if finder_needs_type_condition? && !ignore_default_scope?
-            relation.where(type_condition).create_with(inheritance_column.to_s => sti_name)
+            relation_with_type_condition
           else
-            relation
+            relation_without_type_condition
           end
         end
 
+        def relation_with_type_condition #:nodoc:
```

Comment:
Also we should move those methods to `inheritance.rb` file with `finder_needs_type_condition`